File: JUNGOL/Beginner/#1901 소수 구하기/소수 구하기.cpp
#include <iostream>
#include <vector>
using namespace std;

#define MAX_NUM 1000000

int main() {
	// 에라스토테네스의 체
	vector<bool> isPrime(MAX_NUM+1, true);
	isPrime[1] = false;

	for(int i=4; i<=MAX_NUM; i+=2) isPrime[i] = false;
	for(int i=3; i<=MAX_NUM; i+=2)
		for(int j=i*2; j<=MAX_NUM; j+=i) isPrime[j] = false;

	int n; cin>>n;
	for(int i=0; i<n; i++){
		int m; cin>>m;

		// m이 소수면 바로 출력
		if(isPrime[m]) printf("%d\n", m);
		else{
			// 왼쪽으로 가까운 소수
			int left, right;
			for(int j=m-1; j>0; j--){
				if(isPrime[j]){
					left = j;
					break;
				}
			}

			// 오른쪽으로 가까운 소수
			for(int j=m+1; j<=MAX_NUM; j++){
				if(isPrime[j]){
					right = j;
					break;
				}
			}

			if(m-left == right-m) printf("%d %d\n", left, right);
			else if(m-left > right-m) printf("%d\n", right);
			else printf("%d\n", left);
		}
	}
}

File: JUNGOL/Beginner/#1901 소수 구하기/소수 구하기.py
def main():
    MAX_NUM = 1000000

    isPrime = [True for i in range(MAX_NUM+1)]
    isPrime[1] = False

    for i in range(4, MAX_NUM+1, 2): isPrime[i] = False
    for i in range(3, MAX_NUM+1, 2):
        for j in range(i*2, MAX_NUM+1, i): isPrime[j] = False

    for i in range(int(input())):
        m = int(input())

        if isPrime[m]: print(m)
        else:
            for j in range(m-1, 0, -1):
                if isPrime[j]:
                    left = j
                    break

            for j in range(m+1, MAX_NUM+1):
                if isPrime[j]:
                    right = j
                    break

            if m-left == right-m: print(left, right, sep=' ')
            elif m-left > right-m: print(right)
            else: print(left)

main()
